sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -485,6 +486,42 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case upper @ Aggregate(_, _, lower: Aggregate) if lowerIsRedundant(upper, lower) =>
+      val aliasMap = getAliasMap(lower)
+
+      val newAggregate = upper.copy(
+        child = lower.child,
+        groupingExpressions = upper.groupingExpressions.map(replaceAlias(_, aliasMap)),
+        aggregateExpressions = upper.aggregateExpressions.map(
+          replaceAliasButKeepName(_, aliasMap))
+      )
+
+      // We might have introduces non-deterministic grouping expression
+      PullOutNondeterministic.applyLocally.applyOrElse(newAggregate, identity[LogicalPlan])
```

Comment:
To avoid unnecessary checks, how about this?
```
if (newAggregate.groupingExpressions.exists(!_.deterministic)) {
  PullOutNondeterministic.applyLocally.applyOrElse(newAggregate, identity[LogicalPlan])
}
```
>> // We might have introduces non-deterministic grouping expression

This PR already has tests for this case?

Reply:
Yes, `Remove redundant aggregate with non-deterministic lower` covers this.